Multiple sites

igor47 opened the pull request:
This PR allows us to have multiple versions of the site. Currently, i added an [igor version](https://igor.spaceshipearth.org). To add a new version do the following:

```
inv k8s.namespace <mynamespace>
inv image --namespace=<mynamespace>
inv k8s.run-migrations --namespace=<mynamespace>
```

we should have a `prod` namespace that `www` points at, but i figured we could also create our own versions that we can show to people with minimal overhead or run course-grained experiments with different approaches.

note that this PR changes all the `k8s` `invoke` commands to take an optional `--namespace` argument, you'll often want to pass `prod` or `test` or your own version

igor47 committed:
wrap kubectl interactions in a namespace object

makes sure we always set a namespace correctly when interacting with
kubernetes

igor47 committed:
k8s: initial task for creating namespaces

igor47 committed:
tasks: better namespace creation

we automate some of the annoying stuff around environment creation --
like creating dbs and users and common secrets

igor47 committed:
manage dns when creating namespaces

igor47 committed:
don't overwrite session and mysql secrets

jeredw commented on tasks/utils.py:
```diff
@@ -29,37 +104,7 @@ def load_manifest(mtype, context = {}):
-def random_pod_name():
-  return [p['name'] for p in get_pods() if p['phase'] == 'Running'].pop()
+def random_string(length):
+  return "".join( random.SystemRandom().choices(string.ascii_letters + string.digits, k=length))
```
Same difference, but python 3.6 added a "secrets" module that also exposes SystemRandom and a few more helpers (I guess it is just reassuring that it's the right kind of rng if it is in "secrets"?)

jeredw commented on tasks/k8s.py:
```diff
@@ -1,18 +1,140 @@
+      raise
+
+  # create the namespace
+  with K8SNamespace('default') as ns:
```
Could say DEFAULT_NAMESPACE

jeredw reviewed (approved):
nice, it's cool that this is all automatic

igor47 committed:
treat prod namespace as special

at least, it gets more replicas

igor47 committed:
use prod as source for secrets when initializing namespace

we used to use `default` but this is going out of style now. we can
delete all the secrets and such in `default`
